internal/rollup/rollup.bzl

```diff
@@ -0,0 +1,154 @@
+      if hasattr(dep, "node_sources"):
+        result += dep.node_sources
+  # Note layering: until we have JS interop providers, this needs to know how to
+  # get TypeScript outputs.
+  if hasattr(target, "typescript"):
+    result += target.typescript.es5_sources
+  elif hasattr(target, "files"):
+    result += target.files
+  return struct(node_sources = result)
+
+sources_aspect = aspect(
+    _sources_aspect_impl,
+    attr_aspects = ["deps"],
+)
+
+def _rollup(ctx):
+  rollupConfig = "{0}/rollup.runfiles/build_bazel_rules_nodejs/internal/rollup/rollup.config.js".format(ctx.executable.rollup.dirname)
+  buildFilePath = "/".join(ctx.build_file_path.split("/")[:-1])
+  entryPoint = "bazel-out/host/bin/{0}/rollup.runfiles/{1}/{2}/bazel-out/host/bin/{3}".format(buildFilePath, ctx.workspace_name, buildFilePath, ctx.attr.entry_point)
+
+  args = ["--config", rollupConfig]
+  args += ["--output.file", ctx.outputs.build_es6.path]
+  args += ["--input", entryPoint]
+
+  ctx.action(
+      executable = ctx.executable.rollup,
+      outputs = [ctx.outputs.build_es6],
+      arguments = args
+  )
+
+  argsTS = ["--target", "es5"]
+  argsTS += ["--allowJS"]
+  argsTS += [ctx.outputs.build_es6.path]
+  argsTS += ["--outFile", ctx.outputs.build_es5.path]
+
+  ctx.action(
+      executable = ctx.executable.typescript,
+      inputs = [ctx.outputs.build_es6],
+      outputs = [ctx.outputs.build_es5],
+      arguments = argsTS
+  )
+
+  argsUglify = [ctx.outputs.build_es5.path]
+  argsUglify += ["--output", ctx.outputs.build_es5_min.path]
+
+  ctx.action(
+      executable = ctx.executable.uglify,
+      inputs = [ctx.outputs.build_es5],
+      outputs = [ctx.outputs.build_es5_min],
+      arguments = argsUglify
+  )
+
+  return struct()
+
+rollup = rule(
+    implementation = _rollup,
+    attrs = {
+        "entry_point": attr.string(mandatory=True),
+        "data": attr.label_list(
+            allow_files = True,
+            cfg = "data",
+            aspects=[sources_aspect, module_mappings_runtime_aspect]),
+        "node_modules": attr.label(
+            # By default, binaries use the node_modules in the workspace
+            # where the bazel command is run. This assumes that any needed
+            # dependencies are installed there, commonly due to a transitive
+            # dependency on a package like @bazel/typescript.
+            # See discussion: https://github.com/bazelbuild/rules_typescript/issues/13
+            default = Label("@//:node_modules")),
+        "rollup": attr.label(executable=True, cfg="host", allow_files=True),
+        "typescript": attr.label(executable=True, cfg="host", allow_files=True),
+        "uglify": attr.label(executable=True, cfg="host", allow_files=True)
+    },
+    outputs = {
+        "build_es6": "%{name}.es6.js",
+        "build_es5": "%{name}.js",
+        "build_es5_min": "%{name}.min.js"
+    }
+)
+
+def rollup_macro(data = [], node_modules = Label("@//:node_modules"), **kwargs):
```

Comment:
macros are leaky abstractions - the user will see these individual targets, be able to run them, etc.

instead of a macro, you should be able to have the rule implementation create the multiple actions, chaining outputs from one action as inputs to the next

Reply:
Agreed. First step was to get it working. Will update to chained actions today.